client/client.go

```diff
@@ -0,0 +1,100 @@
+package client
+
+import (
+	"time"
+
+	"golang.org/x/net/context"
+)
+
+// ExistType is used to indicate whether or not a value can be operated on due
+// to the state of the existing node.
+type ExistType uint
+
+var toInitialize = []EntityCollection{}
+
+const (
+	// PrevIgnore ignores any previous content, the default.
+	PrevIgnore = 0
+	// PrevExist only performs the operation if the node already exists.
+	PrevExist = iota
+	// PrevNoExist only performs the operation if the node does not exist.
+	PrevNoExist = iota
+	// PrevValue is used to perform CAS operations. The value must be equivalent if the node exists.
+	PrevValue = iota
+)
+
+// Client is the basic interface to a key-value store with some common properties:
+// * Supports watches
+// * Supports CAS operations
+type Client interface {
+	Get(context.Context, *Pather, *GetOptions) (*Node, error)
+	Set(context.Context, *Pather, []byte, *SetOptions) error
+	Delete(context.Context, *Pather, *DeleteOptions) error
+	Watch(*Pather, chan interface{}, bool, func(*Node, chan interface{}) error) (chan struct{}, chan error)
+	WatchStop(*Pather)
+	Dump(string) (string, error)
+	Path() *Pather
+}
+
+// GetOptions are the retrieve options our abstractions support. These
+// properties will be translated on a per-driver basis to the appropriate
+// idioms.
+type GetOptions struct {
+	Recursive bool
+	Sort      bool // FIXME consul might not have this; is it necessary for anything?
```

Comment:
Ahh, good idea.  We can just reimplement it for stores which lack it as a feature.